src/Microsoft.ML.Transforms/GroupTransform.cs

```diff
@@ -244,52 +242,62 @@ public GroupSchema(Schema inputSchema, IHostEnvironment env, ModelLoadContext ct
+                KeepColumnIndexes = GetColumnIds(inputSchema, _keepColumns, _ectx.Except);
 
-                _columnTypes = BuildColumnTypes(_input, KeepIds);
-                _columnNameMap = BuildColumnNameMap();
-
-                AsSchema = Schema.Create(this);
+                // Compute output schema from the specified input schema.
+                OutputSchema = BuildOutputSchema(inputSchema);
             }
 
-            private Dictionary<string, int> BuildColumnNameMap()
+            /// <summary>
+            /// Compute the output schema of a <see cref="GroupTransform"/> given a input schema.
+            /// </summary>
+            /// <param name="sourceSchema">Input schema.</param>
+            /// <returns>The associated output schema produced by <see cref="GroupTransform"/>.</returns>
+            private Schema BuildOutputSchema(Schema sourceSchema)
             {
-                var map = new Dictionary<string, int>();
-                for (int i = 0; i < _groupCount; i++)
-                    map[_groupColumns[i]] = i;
+                // Create schema build. We will sequentially add group columns and then aggregated columns.
+                var schemaBuilder = new SchemaBuilder();
 
-                for (int i = 0; i < _keepColumns.Length; i++)
-                    map[_keepColumns[i]] = i + _groupCount;
-
-                return map;
-            }
+                // Handle group(-key) columns. Those columns are used as keys to partition rows in the input data; specifically,
+                // rows with the same key value will be merged into one row in the output data.
+                foreach (var groupKeyColumnName in _groupColumns)
+                    schemaBuilder.AddColumn(groupKeyColumnName, sourceSchema[groupKeyColumnName].Type, sourceSchema[groupKeyColumnName].Metadata);
 
-            private static ColumnType[] BuildColumnTypes(Schema input, int[] ids)
-            {
-                var types = new ColumnType[ids.Length];
-                for (int i = 0; i < ids.Length; i++)
+                // Handle aggregated (aka keep) columns.
+                foreach (var groupValueColumnName in _keepColumns)
                 {
-                    var srcType = input[ids[i]].Type;
-                    var primitiveType = srcType as PrimitiveType;
-                    Contracts.Assert(primitiveType != null);
-                    types[i] = new VectorType(primitiveType, size: 0);
+                    // Prepare column's metadata.
+                    var metadataBuilder = new MetadataBuilder();
+                    metadataBuilder.Add(sourceSchema[groupValueColumnName].Metadata,
+                        s => s == MetadataUtils.Kinds.IsNormalized || s == MetadataUtils.Kinds.KeyValues);
+
+                    // Prepare column's type.
+                    var aggragatedValueType = sourceSchema[groupValueColumnName].Type as PrimitiveType;
```

Comment:
>aggragatedValueType [](start = 24, length = 19)

spelling - should be aggregatedValueType

Reply:
Nice catch!

---
In reply to: [245832427](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245832427) [](ancestors = 245832427)